In [81]:
import pandas as pd
from  matplotlib import pyplot as plt
import seaborn as sn
import numpy as np
import warnings
warnings.filterwarnings("ignore")

In [ ]:
data= pd.read_csv("D:\\datasets\\credit-screening.csv")
data.head()

In [2]:
data.shape

(690, 16)

# Preprocessing

In [3]:
#Missing Value
data.isnull().sum()

%X1            0
X2             0
X3             0
X4             0
X5             0
X6             0
X7             0
X8             0
X9             0
X10            0
X11            0
X12            0
X13            0
X14            0
X15            0
class label    0
dtype: int64

# Now we have to convert X1,X4,	X5,	X6,	X7,	X9,	X10,X12,X13 variables into numerical data



In [4]:
#function for coding to character data
#dataframe for charcters and their values
info = pd.DataFrame(columns=['character', 'value'])
def con_num(p,x):
    global info
    j=1
    for i in p.index:    
        data[x]=data[x].replace(to_replace=i, value=j)
        info= info.append({'character': i, 'value': j}, ignore_index=True)
        j+=1
    print("coded Successfully")   
    
    

In [5]:
#Check 	%X1
x1=data.pivot_table(index='%X1',values='class label',aggfunc='count')
x1

class label
%X1             
?             12
a            210
b            468

In [6]:
con_num(x1,'%X1')

coded Successfully


In [7]:
data

%X1     X2      X3 X4 X5  X6  X7    X8 X9 X10  X11 X12 X13    X14  X15  \
0      3  30.83   0.000  u  g   w   v  1.25  t   t    1   f   g  00202    0   
1      2  58.67   4.460  u  g   q   h  3.04  t   t    6   f   g  00043  560   
2      2  24.50   0.500  u  g   q   h  1.50  t   f    0   f   g  00280  824   
3      3  27.83   1.540  u  g   w   v  3.75  t   t    5   t   g  00100    3   
4      3  20.17   5.625  u  g   w   v  1.71  t   f    0   f   s  00120    0   
..   ...    ...     ... .. ..  ..  ..   ... ..  ..  ...  ..  ..    ...  ...   
685    3  21.08  10.085  y  p   e   h  1.25  f   f    0   f   g  00260    0   
686    2  22.67   0.750  u  g   c   v  2.00  f   t    2   t   g  00200  394   
687    2  25.25  13.500  y  p  ff  ff  2.00  f   t    1   t   g  00200    1   
688    3  17.92   0.205  u  g  aa   v  0.04  f   f    0   f   g  00280  750   
689    3  35.00   3.375  u  g   c   h  8.29  f   f    0   t   g  00000    0   

    class label  
0             +  
1             +  
2             +  
3             +  
4             +  
..          ...  
685           -  
686           -  
687           -  
688           -  
689           -  

[690 rows x 16 columns]

In [8]:
#Check X4 
x4=data.pivot_table(index='X4',values='class label',aggfunc='count')
x4

class label
X4             
?             6
l             2
u           519
y           163

In [9]:
#Now replace these values with numerical values ?=1, l=2 ,u=3,y=4
con_num(x4,'X4')

coded Successfully


In [10]:
info

character value
0         ?     1
1         a     2
2         b     3
3         ?     1
4         l     2
5         u     3
6         y     4

In [11]:
#check X5
x5=data.pivot_table(index='X5',values='class label',aggfunc='count')
x5

class label
X5             
?             6
g           519
gg            2
p           163

In [12]:
#Now replace these values with numerical values ?=1, g=2 ,gg=3,p=4
con_num(x5,'X5')

coded Successfully


In [13]:
#chexk X6
x6=data.pivot_table(index='X6',values='class label',aggfunc='count')
x6

class label
X6             
?             9
aa           54
c           137
cc           41
d            30
e            25
ff           53
i            59
j            10
k            51
m            38
q            78
r             3
w            64
x            38

In [14]:
#Now replace these values with numerical values ?=1, g=2 ,gg=3,p=4
con_num(x6,'X6')

coded Successfully


In [15]:
#check X7
x7=data.pivot_table(index='X7', values='class label', aggfunc='count')
x7

class label
X7             
?             9
bb           59
dd            6
ff           57
h           138
j             8
n             4
o             2
v           399
z             8

In [16]:
con_num(x7,'X7')


coded Successfully


In [17]:
#check X9
x9=data.pivot_table(index='X9', values='class label', aggfunc='count')
x9

class label
X9             
f           329
t           361

In [18]:
con_num(x9,'X9')

coded Successfully


In [19]:
#check X10
x10=data.pivot_table(index='X10', values='class label', aggfunc='count')
x10

class label
X10             
f            395
t            295

In [20]:
con_num(x10,'X10')

coded Successfully


In [21]:
#chexk X12
x12=data.pivot_table(index='X12', values='class label', aggfunc='count')
x12

class label
X12             
f            374
t            316

In [22]:
con_num(x12,'X12')

coded Successfully


In [23]:
#Check X13
x13=data.pivot_table(index='X13', values='class label', aggfunc='count')
x13

class label
X13             
g            625
p              8
s             57

In [24]:
con_num(x13,'X13')

coded Successfully


In [25]:
#Data after coded into numercical form
data.head()

%X1     X2     X3  X4  X5  X6  X7    X8  X9  X10  X11  X12  X13    X14  \
0    3  30.83  0.000   3   2  14   9  1.25   2    2    1    1    1  00202   
1    2  58.67  4.460   3   2  12   5  3.04   2    2    6    1    1  00043   
2    2  24.50  0.500   3   2  12   5  1.50   2    1    0    1    1  00280   
3    3  27.83  1.540   3   2  14   9  3.75   2    2    5    2    1  00100   
4    3  20.17  5.625   3   2  14   9  1.71   2    1    0    1    3  00120   

   X15 class label  
0    0           +  
1  560           +  
2  824           +  
3    3           +  
4    0           +

# check in any other characters in X2,X3,X8,x14,X15

In [26]:
from string import punctuation
punctuation

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'

In [27]:
#Function fo checking special characters
spc_chars= [] 
def chk_symbol(x):    
    global dict,spc_chars
    status=0
    sym='!"#$%&\'()*,./:;<=>?@[\\]^_`{|}~'
    flag=0
    for s in sym:
        flag=0 
        i=0
        for d in data[x]:
            if d==s:
                spc_chars.append(i)
                flag=1
            i+=1  
        if flag==1:         
          status=1
    if status==1:
        print("Special charctes found at indexes")
        print(spc_chars)
    else:
        print("Special charctes not found ")

In [28]:
#check for X2
chk_symbol('X2')

Special charctes found at indexes
[83, 86, 92, 97, 254, 286, 329, 445, 450, 500, 515, 608]


In [29]:
#check for X8
chk_symbol('X8')

Special charctes not found 


In [30]:
#check for X14
chk_symbol('X14') 

Special charctes found at indexes
[83, 86, 92, 97, 254, 286, 329, 445, 450, 500, 515, 608, 71, 202, 206, 243, 270, 278, 330, 406, 445, 456, 592, 622, 626]


In [31]:
#check for X15
chk_symbol('X15') 

Special charctes not found 


In [32]:
#Special charcters found at folowing indexes
print(spc_chars,'no of rows =',len(spc_chars))

[83, 86, 92, 97, 254, 286, 329, 445, 450, 500, 515, 608, 71, 202, 206, 243, 270, 278, 330, 406, 445, 456, 592, 622, 626] no of rows = 25


In [33]:
#Remove this rows
data=data.drop(spc_chars)

In [34]:
data.shape

(666, 16)

In [35]:
#coded data
info

character value
0          ?     1
1          a     2
2          b     3
3          ?     1
4          l     2
5          u     3
6          y     4
7          ?     1
8          g     2
9         gg     3
10         p     4
11         ?     1
12        aa     2
13         c     3
14        cc     4
15         d     5
16         e     6
17        ff     7
18         i     8
19         j     9
20         k    10
21         m    11
22         q    12
23         r    13
24         w    14
25         x    15
26         ?     1
27        bb     2
28        dd     3
29        ff     4
30         h     5
31         j     6
32         n     7
33         o     8
34         v     9
35         z    10
36         f     1
37         t     2
38         f     1
39         t     2
40         f     1
41         t     2
42         g     1
43         p     2
44         s     3

In [36]:
X=data.iloc[:,0:-1]
y=data.iloc[:,-1]

In [37]:
#Model Training
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size =0.2, random_state =1)

# Binary Logistic Regression

In [38]:
from sklearn.linear_model import LogisticRegression
LR=LogisticRegression()
LR.fit(X_train,y_train)

C:\Users\reshw\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=None, solver='warn', tol=0.0001, verbose=0,
                   warm_start=False)

In [39]:
#Predict values
y_pred = LR.predict(X_test)
y_pred

array(['-', '-', '+', '+', '-', '-', '-', '+', '-', '+', '+', '+', '-',
       '-', '+', '+', '+', '+', '+', '+', '-', '+', '+', '+', '-', '-',
       '-', '+', '-', '-', '-', '-', '+', '+', '-', '-', '+', '+', '-',
       '-', '-', '-', '+', '+', '+', '-', '-', '+', '-', '+', '+', '-',
       '+', '+', '-', '+', '+', '+', '-', '+', '-', '+', '-', '+', '+',
       '-', '-', '-', '-', '-', '+', '-', '-', '+', '+', '-', '-', '+',
       '-', '-', '-', '-', '-', '+', '-', '+', '+', '+', '+', '+', '-',
       '-', '+', '+', '+', '+', '+', '-', '-', '-', '+', '-', '-', '+',
       '+', '+', '-', '+', '-', '-', '+', '-', '-', '+', '-', '-', '-',
       '-', '-', '-', '+', '+', '-', '-', '-', '-', '-', '-', '-', '+',
       '-', '+', '-', '-'], dtype=object)

In [40]:
#confusion matrix
from sklearn import metrics
cnf_mat=metrics.confusion_matrix(y_test,y_pred)
print(cnf_mat)

[[51  7]
 [10 66]]


In [41]:
print('Accuracy',metrics.accuracy_score(y_test,y_pred))

Accuracy 0.8731343283582089


# Ordinal Logistic Regression

In [68]:
from sklearn.linear_model import LogisticRegression
olr = LogisticRegression(solver = 'lbfgs')
olr.fit(X_train,y_train)
y_pred = LR.predict(X_test)
print('Accuracy',metrics.accuracy_score(y_test,y_pred))

Accuracy 0.8731343283582089


C:\Users\reshw\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:947: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.
  "of iterations.", ConvergenceWarning)


# Multi Class Logistic Regression

In [42]:
from sklearn.linear_model import LogisticRegression

clf = LogisticRegression(random_state=0, multi_class='multinomial', solver='newton-cg')
# Train model
clf.fit(X_train,y_train)
y_pred = clf.predict(X_test)

print('Accuracy thru Multiclass Logistic Regression ', metrics.accuracy_score(y_test, y_pred))

Accuracy thru Multiclass Logistic Regression  0.8507462686567164


In [43]:
#confusion matrix
from sklearn import metrics
cnf_mat=metrics.confusion_matrix(y_test,y_pred)
print(cnf_mat)

[[49  9]
 [11 65]]


# QuadraticDiscriminantAnalysis


In [44]:
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis
QDA=QuadraticDiscriminantAnalysis()
QDA.fit(X_train,y_train)

C:\Users\reshw\Anaconda3\lib\site-packages\sklearn\discriminant_analysis.py:693: UserWarning: Variables are collinear
  warnings.warn("Variables are collinear")


QuadraticDiscriminantAnalysis(priors=None, reg_param=0.0,
                              store_covariance=False, tol=0.0001)

In [45]:
y_pred = QDA.predict(X_test)

print('Accuracy thru Multiclass Logistic Regression ', metrics.accuracy_score(y_test, y_pred))

Accuracy thru Multiclass Logistic Regression  0.5522388059701493


In [46]:
#confusion matrix
from sklearn import metrics
cnf_mat=metrics.confusion_matrix(y_test,y_pred)
print(cnf_mat)

[[51  7]
 [53 23]]


# Linear Discriminant Analysis

In [47]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
LDM=LinearDiscriminantAnalysis()
LDM.fit(X_train,y_train)

LinearDiscriminantAnalysis(n_components=None, priors=None, shrinkage=None,
                           solver='svd', store_covariance=False, tol=0.0001)

In [48]:
y_pred = QDA.predict(X_test)

print('Accuracy thru Multiclass Logistic Regression ', metrics.accuracy_score(y_test, y_pred))

Accuracy thru Multiclass Logistic Regression  0.5522388059701493


In [49]:
#confusion matrix
from sklearn import metrics
cnf_mat=metrics.confusion_matrix(y_test,y_pred)
print(cnf_mat)

[[51  7]
 [53 23]]


# KNN

In [50]:
from sklearn.neighbors import KNeighborsClassifier
KNN=KNeighborsClassifier()
KNN.fit(X_train,y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=5, p=2,
                     weights='uniform')

In [51]:
#Predict values
y_pred = LR.predict(X_test)
print('Accuracy',metrics.accuracy_score(y_test,y_pred))

Accuracy 0.8731343283582089


In [52]:
#confusion matrix
from sklearn import metrics
cnf_mat=metrics.confusion_matrix(y_test,y_pred)
print(cnf_mat)

[[51  7]
 [10 66]]


# NaiveBayes


In [53]:
from sklearn.naive_bayes import GaussianNB,MultinomialNB
GNB=GaussianNB()
GNB.fit(X_train,y_train)

GaussianNB(priors=None, var_smoothing=1e-09)

In [54]:
y_pred =GNB.predict(X_test)
print('Accuracy',metrics.accuracy_score(y_test,y_pred))

Accuracy 0.7686567164179104


In [55]:
#confusion matrix
from sklearn import metrics
cnf_mat=metrics.confusion_matrix(y_test,y_pred)
print(cnf_mat)

[[35 23]
 [ 8 68]]


# Decision Tree

In [58]:
from sklearn.tree import DecisionTreeClassifier
DTC=DecisionTreeClassifier()
DTC.fit(X_train,y_train)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
                       max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort=False,
                       random_state=None, splitter='best')

In [60]:
y_pred =DTC.predict(X_test)
print('Accuracy',metrics.accuracy_score(y_test,y_pred))

Accuracy 0.8208955223880597


In [62]:
#confusion matrix
from sklearn import metrics
cnf_mat=metrics.confusion_matrix(y_test,y_pred)
print(cnf_mat)

[[49  9]
 [15 61]]


# Random Forest

In [65]:
from sklearn.ensemble import RandomForestClassifier

regressor = RandomForestClassifier(n_estimators=20, random_state=0)
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)
print('Accuracy',metrics.accuracy_score(y_test,y_pred))

Accuracy 0.8582089552238806


# Gradient Boost

In [66]:
from sklearn.ensemble import GradientBoostingClassifier
GBC=GradientBoostingClassifier()
GBC.fit(X_train,y_train)
y_pred = GBC.predict(X_test)
print('Accuracy',metrics.accuracy_score(y_test,y_pred))

Accuracy 0.8582089552238806


In all the models Binary Logisting Regression ,Ordinal Logistic REgression ,KNN are give 87.31 % accuracy

In [84]:
#So i have to save model
from sklearn.externals import joblib
joblib.dump(KNN ,"Credit_Card_Approval_KNN.pkl")

['Credit_Card_Approval_KNN.pkl']

In [85]:
#Load Saved Model from joblib
CCA=joblib.load("Credit_Card_Approval_KNN.pkl")
CCA

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=5, p=2,
                     weights='uniform')

In [69]:
data.head()

%X1     X2     X3  X4  X5  X6  X7    X8  X9  X10  X11  X12  X13    X14  \
0    3  30.83  0.000   3   2  14   9  1.25   2    2    1    1    1  00202   
1    2  58.67  4.460   3   2  12   5  3.04   2    2    6    1    1  00043   
2    2  24.50  0.500   3   2  12   5  1.50   2    1    0    1    1  00280   
3    3  27.83  1.540   3   2  14   9  3.75   2    2    5    2    1  00100   
4    3  20.17  5.625   3   2  14   9  1.71   2    1    0    1    3  00120   

   X15 class label  
0    0           +  
1  560           +  
2  824           +  
3    3           +  
4    0           +

In [86]:
#Testing
X1= 2
X2=60.00
X3= 2
X4= 3
X5=1
X6= 8
X7= 9
X8 =1
X9 =2
X10 =2
X11=4
X12=1
X13=2 
X14='00043'
X15=200
input=[X1,X2,X3,X4,X5,X6,X7,X8,X9,X10,X11,X12,X13,X14,X15]
outcome=CCA.predict([input])
if outcome=='+':
    print("Creditcard approved")
else:
    print("Creditcard not approved")

Creditcard not approved


In [90]:
input=[3,30.83,0,3,2,14,9,1.25,2,2,1,1,1,202,0]
outcome=CCA.predict([input])
if outcome=='+':
    print("Creditcard approved")
else:
    print("Creditcard not approved")

Creditcard not approved
